src/http_interface.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The haproxy http interface module."""
+
+import json
+import logging
+
+from ops import RelationBrokenEvent, RelationChangedEvent, RelationJoinedEvent
+from ops.charm import CharmBase, CharmEvents, RelationEvent
+from ops.framework import EventSource, Object
+from ops.model import ModelError, Relation, RelationDataContent
+
+import legacy
+
+logger = logging.getLogger()
+SERVICES_CONFIGURATION_KEY = "services"
+
+
+class HTTPDataProvidedEvent(RelationEvent):
+    """Event representing that http data has been provided."""
+
+
+class HTTPDataRemovedEvent(RelationEvent):
+    """Event representing that http data has been removed."""
+
+
+class HTTPProviderEvents(CharmEvents):
+    """Container for HTTP Provider events.
+
+    Attrs:
+        data_provided: Custom event when integration data is provided.
+        data_removed: Custom event when integration data is removed.
+    """
+
+    data_provided = EventSource(HTTPDataProvidedEvent)
+    data_removed = EventSource(HTTPDataRemovedEvent)
+
+
+class _IntegrationInterfaceBaseClass(Object):
```

Comment:
This seems like an abstract base class, is that needed in this case? It looks like this is only used as the base class for one other class?

Reply:
I have the HTTPRequirer class also planned for the `website` relation, but that has not been implemented yet.